plugins/javaembed/javaembed.cpp

```diff
@@ -1508,589 +1509,79 @@ class JavaThreadContext
-                break;
-            }
-        }
-        __isAllResult = false;
-        __resultBytes = elemSize == UNKNOWN_LENGTH ? outBytes : elemSize * numResults;
-        __result = out.detachdata();
+        jobject systemClassLoaderObj = JNIenv->CallStaticObjectMethod(javaLangClassLoaderClass, getSystemClassLoaderMethod);
+        checkException();
+        assertex(systemClassLoaderObj);
+        return systemClassLoaderObj;
     }
-    inline const char *querySignature()
+
+    void setThreadClassLoader(jobject classLoader)
     {
-        return argsig.get();
+        JNIenv->ExceptionClear();
+        jclass javaLangThreadClass = JNIenv->FindClass("java/lang/Thread");
+        checkException();
+        jmethodID currentThreadMethod = JNIenv->GetStaticMethodID(javaLangThreadClass, "currentThread", "()Ljava/lang/Thread;");
+        checkException();
+        jobject threadObj = JNIenv->CallStaticObjectMethod(javaLangThreadClass, currentThreadMethod);
+        checkException();
+        jmethodID setContextClassLoaderMethod = JNIenv->GetMethodID(javaLangThreadClass, "setContextClassLoader", "(Ljava/lang/ClassLoader;)V");
+        checkException();
+        JNIenv->CallObjectMethod(threadObj, setContextClassLoaderMethod, classLoader);
+        checkException();
     }
-    size32_t getRowResult(jobject result, ARowBuilder &builder)
+
+    jobject getThreadClassLoader()
     {
-        const RtlTypeInfo *typeInfo = builder.queryAllocator()->queryOutputMeta()->queryTypeInfo();
-        assertex(typeInfo);
-        RtlFieldStrInfo dummyField("<row>", NULL, typeInfo);
-        JavaRowBuilder javaRowBuilder(JNIenv, &dummyField, result);
-        return typeInfo->build(builder, 0, &dummyField, javaRowBuilder);
+        JNIenv->ExceptionClear();
+        jclass javaLangThreadClass = JNIenv->FindClass("java/lang/Thread");
+        checkException();
+        jmethodID currentThreadMethod = JNIenv->GetStaticMethodID(javaLangThreadClass, "currentThread", "()Ljava/lang/Thread;");
+        checkException();
+        jobject threadObj = JNIenv->CallStaticObjectMethod(javaLangThreadClass, currentThreadMethod);
+        checkException();
+        jmethodID getContextClassLoaderMethod = JNIenv->GetMethodID(javaLangThreadClass, "getContextClassLoader", "()Ljava/lang/ClassLoader;");
+        checkException();
+        jobject contextClassLoaderObj = JNIenv->CallObjectMethod(threadObj, getContextClassLoaderMethod);
+        checkException();
+        assertex(contextClassLoaderObj);
+        return contextClassLoaderObj;
     }
+
     void writeObjectResult(jobject result, IEsdlDefinition *esdl, const char *esdlservice, const char *name, IXmlWriter *writer)
     {
         JavaObjectXmlWriter x(JNIenv, result, name, *esdl, esdlservice, *writer);
         x.write();
     }
-
+    void registerContext(IEmbedFunctionContext *ctx)
+    {
+        contexts.append(*ctx);
```

Comment:
Worth commenting that there is no need for a critical section because there is an instance for each thread.

Reply:
done